tensorly/tenalg/proximal.py

```diff
@@ -248,3 +248,196 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+    s = tl.zeros(tl.shape(x_vec), **tl.context(x_vec))
+
+    for iteration in range(n_iter_max):
+
+        if iteration > 0 or tl.all(x_vec == 0):
+            indice = tl.argmax(x_gradient)
+            passive_set = tl.index_update(passive_set, tl.index[indice], True)
+            active_set = tl.index_update(active_set, tl.index[indice], False)
+        # To avoid singularity error when initial x exists
+        try:
+            a = tl.tensor(tl.to_numpy(UtU)[passive_set, :][:, passive_set], **tl.context(UtU))
+            c = tl.solve(a, Utm[passive_set])
+            indice_list = []
+            for i in range(tl.shape(s)[0]):
+                if passive_set[i]:
+                    indice_list.append(i)
+                    s = tl.index_update(s, tl.index[int(i)], c[len(indice_list) - 1])
+                else:
+                    s = tl.index_update(s, tl.index[int(i)], 0)
+        # Start from zeros if solve is not achieved  
+        except:
+            x_vec = tl.zeros(tl.shape(UtU)[1])
+            s = tl.zeros(tl.shape(x_vec), **tl.context(x_vec))
+            passive_set = x_vec > 0
+            active_set = x_vec <= 0
+            if True in active_set:
+                indice = tl.argmax(x_gradient)
+                passive_set = tl.index_update(passive_set, tl.index[indice], True)
+                active_set = tl.index_update(active_set, tl.index[indice], False)
+            a = tl.tensor(tl.to_numpy(UtU)[passive_set, :][:, passive_set], **tl.context(UtU))
+            c = tl.solve(a, Utm[passive_set])
+            indice_list = []
+            for i in range(tl.shape(s)[0]):
+                if passive_set[i]:
+                    indice_list.append(i)
+                    s = tl.index_update(s, tl.index[int(i)], c[len(indice_list) - 1])
+                else:
+                    s = tl.index_update(s, tl.index[int(i)], 0)
+
+        # update support vector if it is necessary
+        if tl.min(s[passive_set]) <= 0:
+            for i in range(len(passive_set)):
+                alpha = tl.min(x_vec[passive_set][s[passive_set] <= 0] / (x_vec[passive_set][s[passive_set] <= 0] - s[passive_set][s[passive_set] <= 0]))
+                update = alpha * (s - x_vec)
+                x_vec = x_vec + update
+                passive_set = x_vec > 0
+                active_set = x_vec <= 0
+                a = tl.tensor(tl.to_numpy(UtU)[passive_set, :][:, passive_set], **tl.context(UtU))
+                c = tl.solve(a, Utm[passive_set])
+                indice_list = []
+                for i in range(tl.shape(s)[0]):
+                    if passive_set[i]:
+                        indice_list.append(i)
+                        s = tl.index_update(s, tl.index[int(i)], c[len(indice_list) - 1])
+                    else:
+                        s = tl.index_update(s, tl.index[int(i)], 0)
+
+                if True not in passive_set or tl.min(s[passive_set]) > 0:
+                    break
+        # set x to s
```

Comment:
Could we also find a more explicit name for `s` (and `c` and `a`?) 
The other names we now use make the code much easier to read (passive/active_set, x_gradient, etc)!

Reply:
I changed c to passive_solution, s to support_vec and removed a by merging that calculation with the next line.